google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java

```diff
@@ -321,50 +348,385 @@ private boolean isAlive() {
                         .setSubscription(subscription)
-                        .addAllAckIds(idChunk)
+                        .addAllAckIds(ackIdsInRequest)
                         .setAckDeadlineSeconds(modack.deadlineExtensionSeconds)
                         .build());
         ApiFutures.addCallback(future, loggingCallback, directExecutor());
         pendingOperations++;
       }
     }
+    ackOperationsWaiter.incrementPendingCount(pendingOperations);
+    return futureTable;
+  }
 
-    for (List<String> idChunk : Lists.partition(acksToSend, MAX_PER_REQUEST_CHANGES)) {
+  private ModacksToRetryModacksToFail processModackFutures(
+      Table<Integer, ImmutableList<String>, SettableApiFuture<Map<String, String>>>
+          futureCallbackTable) {
+    List<PendingModifyAckDeadline> modacksToRetry = new ArrayList<>();
+    List<String> modackIdsFailed = new ArrayList<>();
+    for (Integer deadlineExtensionSeconds : futureCallbackTable.rowKeySet()) {
+      Map<ImmutableList<String>, SettableApiFuture<Map<String, String>>> rowMap =
+          futureCallbackTable.row(deadlineExtensionSeconds);
+      List<String> ackIdsToRetry = new ArrayList<>();
+      List<String> ackIdsToFail = new ArrayList<>();
+      for (ImmutableList<String> ackIdsInRequest : rowMap.keySet()) {
+        try {
+          Map<String, String> metadataMap = rowMap.get(ackIdsInRequest).get();
+          metadataMap.forEach(
+              (ackId, errorMessage) -> {
+                if (errorMessage.startsWith(TRANSIENT_ERROR_METADATA_PREFIX)) {
+                  ackIdsToRetry.add(ackId);
+                  logger.log(Level.WARNING, "Transient error detected, will retry", errorMessage);
+                } else if (errorMessage.startsWith(PERMANENT_ERROR_METADATA_PREFIX)) {
+                  ackIdsToFail.add(ackId);
+                  logger.log(
+                      Level.WARNING, "Permanent error detected, will not retry", errorMessage);
+                }
+              });
+        } catch (Throwable t) {
+          logger.log(Level.WARNING, "Failed to retrieve future. resending ackIds", t);
+          ackIdsToRetry.addAll(ackIdsInRequest);
+        }
+      }
+      modacksToRetry.add(new PendingModifyAckDeadline(deadlineExtensionSeconds, ackIdsToRetry));
+      modackIdsFailed.addAll(ackIdsToFail);
+    }
+    return new ModacksToRetryModacksToFail(modacksToRetry, modackIdsFailed);
+  }
+
+  private void sendAcksWithFutures(
+      List<MessageDispatcher.AckWithMessageFuture> ackIdWithMessageFutureToSend) {
+    if (!ackIdWithMessageFutureToSend.isEmpty()) {
+      Map<String, SettableApiFuture<Map<String, String>>> ackFutureMap =
+          sendUnaryAcks(ackIdWithMessageFutureToSend);
+
+      if (ackFutureMap.isEmpty()) {
+        return;
+      }
+      List<MessageDispatcher.AckWithMessageFuture> acksToResend =
+          processAckFutures(ackIdWithMessageFutureToSend, ackFutureMap);
+      sendAcksWithFutures(acksToResend);
```

Comment:
This is happening immediately. Should there be exponential back off?

Reply:
This has been updated to use exponential backoff